### PR TITLE
Merge join inclusion correctness

### PR DESCRIPTION
**Pull request by max-hoffman**
Merge join was not principled selecting 1) monotonic filters and 2) sorting indexes that aligned with the filter of choice. We also failed to push filters referencing a single table out of join conditions.

### CODE REVIEW

File: sql/analyzer/indexed_joins.go [hunk 1]
```diff
@@ -432,24 +434,56 @@ func collectJoinConds(attributeSource string, filters ...sql.Expression) []*join
 	return conds
 }
 
+func collectMergeJoinConds(la, ra string, filters ...sql.Expression) ([]*joinColExpr, []*joinColExpr) {
```
Comment (fulghum) [inline sql/analyzer/indexed_joins.go:437]: This function doesn't seem to be called from anywhere. Is this just dead code?

**Review by fulghum — APPROVED** [review]
Code generally looks good. I just had a couple suggestions and a couple questions to make sure I understood how the code is supposed to work.

File: enginetest/queries/query_plans.go [hunk 2]
```diff
@@ -17110,38 +17111,39 @@ WHERE
...
-			"             │                           │                       ├─ static: [{[NULL, ∞)}]\n" +
-			"             │                           │                       └─ Table\n" +
-			"             │                           │                           └─ name: EPZU6\n" +
+			"             │                           │               └─ HashLookup\n" +
+			"             │                           │                   ├─ source: TUPLE(0 (tinyint), uct.FTQLQ:1, uct.ZH72S:2, uct.LJLUM:5)\n" +
+			"             │                           │                   ├─ target: TUPLE(I7HCR.SWCQV:5!null, I7HCR.TOFPN:1!null, I7HCR.SJYN2:2!null, I7HCR.BTXC5:3!null)\n" +
+			"             │                           │                   └─ CachedResults\n" +
+			"             │                           │                       └─ TableAlias(I7HCR)\n" +
+			"             │                           │                           └─ Table\n" +
+			"             │                           │                               └─ name: EPZU6\n" +
```
[the remark this answers is not in the document]
Comment (max-hoffman) [reply enginetest/queries/query_plans.go:17146]: the main problem was correctness, the merge join was not valid. It is probably slower, and it is also possible there is a preferable join strategy. I need to review your point about sort selection being too strict, it's possible we can still merge join here. i did not notice a significant performance impact here introducing the hash join, however.

[the remark this answers is not in the document]
Comment (fulghum) [reply enginetest/queries/query_plans.go:17146]: cool cool, makes sense to me

File: sql/analyzer/indexed_joins.go [hunk 3]
```diff
@@ -598,92 +634,145 @@ func addMergeJoins(m *Memo) error {
...
+
+func isMonotonic(e sql.Expression) bool {
+	switch e := e.(type) {
+	case *expression.Arithmetic:
+		if e.Op == sqlparser.MinusStr {
+			// TODO minus can be OK if it's not on the GetField
+			return false
+		}
+	case expression.Comparer, *expression.Literal, *expression.GetField, *expression.And,
+		*expression.Tuple, *expression.IsNull, *expression.BindVar:
```
[the remark this answers is not in the document]
Comment (max-hoffman) [reply sql/analyzer/indexed_joins.go:758]: For that example, as long as the get fields are not negated, the value of the expression will be strictly increasing (monotonic). So a big negative literal in the expression value will still increase as we move along the index, as long as we are always adding x. On closer inspection, I should maybe take out the `InTuple`,`Regex`, and maybe `GreaterThan` Comparer types, which might decrease as x increases.

File: enginetest/queries/query_plans.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (max-hoffman) [reply enginetest/queries/query_plans.go]: i did manage to claw back the merge joins with the tighter correctness guarantees FYI

**Pull request merged**